### PR TITLE
Add Support for `primary_entity` in Semantic Models 

### LINKED ISSUE #694: [Feature] Add Support for `primary_entity` in Semantic Models
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing metricflow functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Following the update to DSI (https://github.com/dbt-labs/dbt-semantic-interfaces/issues/123), MF should be updated to support querying based on the primary_entity.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by plypaul**
Resolves #694 

### Description

Please see the linked issue for more details.

### CODE REVIEW

**Review by tlento — COMMENTED** [review]
Something's wonky with time dimension handling.

1. Time dimensions don't appear to require the primary entity specification in all cases
2. We can apparently have duplicate `primary_entity__time_dimension_name` values across semantic models in the semantic manifest (take a look at the .yaml files for the extended_date_model - with this PR there are now multiple `booking_id__ds` dimensions)
3. We appear to no longer list time dimensions as linkable elements in certain cases, such as getting linkable specs for a metric

Is all of this expected? If so, what does this look like for an end user today? When they run the equivalent of `mf list-dimensions` would things like `ds_partitioned` no longer show up (as opposed to showing up as `booking_id__ds_partitioned` or whatever)? When someone requests a query grouped by `booking_id__ds_partitioned`, will that work? Can users define semantic models with only date dimensions and no primary entity and have those pass validation?

I'm also concerned that we have an apparently illegal model spec that's passing validation in the updated extended_date_model yaml files. Am I misreading something there?

**Comment by plypaul** [comment]
> Time dimensions don't appear to require the primary entity specification in all cases.

Explained in the inline comment, but it was missing a call to a validation.

> We can apparently have duplicate primary_entity__time_dimension_name values across semantic models in the semantic manifest (take a look at the .yaml files for the extended_date_model - with this PR there are now multiple booking_id__ds dimensions)

Apparently, validations don't get run on the semantic manifests defined in the test repo anymore. I'm going to guess this was an oversight when validations were ripped out of MF and put into DSI. I'll make an update to add this one + fix the model.

> We appear to no longer list time dimensions as linkable elements in certain cases, such as getting linkable specs for a metric

Responded in a inline comment.

> Is all of this expected? If so, what does this look like for an end user today? When they run the equivalent of mf list-dimensions would things like ds_partitioned no longer show up (as opposed to showing up as booking_id__ds_partitioned or whatever)?

It should show up as `booking_id__ds_partitioned` if it's a valid item to group by.

> When someone requests a query grouped by booking_id__ds_partitioned, will that work?

Yes, that will work. e.g. https://github.com/dbt-labs/metricflow/pull/695/commits/d2b76ae70f0e7f5df83135e4a3ae011b0ad69f71#diff-3c12c993132bb029b46926fcdb6ad2c7cf1a028f46e283c9a7b0999f4599ce60R7

>Can users define semantic models with only date dimensions and no primary entity and have those pass validation?
Right now the rule for the semantic models is that if it contains dimensions, it requires a primary entity. So that would fail the validation. Have thoughts on that condition?

**Review by tlento — APPROVED** [review]
Ok, I think what we have makes sense for what we've built, but it is a bit more limiting than expected.

We'll want to find a way to unlock joins across dimensions (or entities) that are technically not the same, but are nevertheless joinable, but I think that's a bit more work than we can afford to do at this time. End users will have a work-around for this via dbt if they really need to do it before we can get the update out, which is good since there's a good chance we won't be able to make it broadly available until dbt-core 1.7.

Regarding this:

> Right now the rule for the semantic models is that if it contains dimensions, it requires a primary entity. So that would fail the validation. Have thoughts on that condition?

I think that's the right starting point. We may want to relax this constraint, and if so I think the right way to relax it will be via whatever mechanism we use to allow for joins across primary entity boundaries. But I also think we won't have any need to relax this constraint if we allow people to specify that dimensions x and y can be joined even if the qualified names don't match.

File: metricflow/model/semantics/semantic_model_lookup.py [hunk 1]
```diff
@@ -190,14 +193,45 @@ def _add_semantic_model(self, semantic_model: SemanticModel) -> None:
...
+                raise RuntimeError(
+                    f"Aggregation time dimension does not have a time granularity set: {agg_time_dimension}"
+                )
+
+            primary_entity = SemanticModelLookup._resolved_primary_entity(semantic_model)
+
+            if primary_entity is None:
+                raise RuntimeError(
+                    f"The semantic model should have a primary entity since there are measures, but it does not. "
+                    f"Semantic model is:\n{pformat_big_objects(semantic_model)}"
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/model/semantics/semantic_model_lookup.py]: Fixed.

File: metricflow/model/semantics/semantic_model_lookup.py [hunk 2]
```diff
@@ -246,3 +280,43 @@ def get_entity_from_semantic_model(
...
+        entities_with_type_primary = tuple(
+            entity for entity in semantic_model.entities if entity.type == EntityType.PRIMARY
+        )
+
+        # This should be caught by the validation, but adding sanity check.
+        assert len(entities_with_type_primary) <= 1, f"Found >1 primary entity in {semantic_model}"
+        if primary_entity_reference is not None:
+            assert len(entities_with_type_primary) == 0, (
+                f"The primary_time field was set to {primary_entity_reference}, but there are non-zero entities with "
+                f"type {EntityType.PRIMARY} in {semantic_model}"
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/model/semantics/semantic_model_lookup.py]: Fixed.

**Comment by plypaul** [comment]
Seems like there's an issue with the permissions for the CI user in Snowflake that's prevent it from running commands to drop / create schemas. Going to merge for now - will check those permissions with the Snowflake admin.

**Pull request merged**